src/Store/CompletionStoreConnector.re

```diff
@@ -11,139 +11,192 @@ open Actions;
-           | None => ()
-           | Some(buffer) =>
-             let cursorPosition = Editor.getPrimaryCursor(ed);
-             let meetOpt =
-               CompletionMeet.createFromBufferLocation(
-                 ~location=cursorPosition,
-                 buffer,
-               );
-             switch (meetOpt) {
-             | None =>
-               lastMeet := defaultMeet;
-               dispatch(Actions.CompletionEnd);
-             | Some(meet) =>
-               open CompletionMeet;
-               let {line, column}: Location.t = getLocation(meet);
-               let base = getBase(meet);
-               // Check if our 'meet' position has changed
-               let newMeet = {
-                 completionMeetBufferId: bufferId,
-                 completionMeetLine: line,
-                 completionMeetColumn: column,
-               };
-               if (!equals(~line, ~column, ~bufferId, lastMeet^)) {
-                 Log.infof(m =>
-                   m(
-                     "New completion meet: %s",
-                     CompletionMeet.toString(meet),
-                   )
-                 );
-                 dispatch(CompletionStart(meet));
-                 dispatch(CompletionBaseChanged(base));
-               } else if (!String.equal(base, lastBase^)) {
-                 // If we're at the same position... but our base is different...
-                 // fire a base change
-                 lastBase := base;
-                 dispatch(CompletionBaseChanged(base));
-                 Log.info("New completion base: " ++ base);
-               };
-               lastMeet := newMeet;
-             };
-           };
-         });
-    });
+  let update = (f, state: State.t) => (
+    {...state, completions: f(state.completions)},
+    Isolinear.Effect.none,
+  );
+};
 
-  let updater = (state: State.t, action: Actions.t) => {
-    switch (action) {
-    | Command("acceptSelectedSuggestion") => (state, applyCompletion(state))
-    | ChangeMode(mode) when mode == Vim.Types.Insert => (
+let start = () => {
+  let updater = (state: State.t) =>
+    fun
+    | Command("acceptSelectedSuggestion") => Actions.applyCompletion(state)
+
+    | ChangeMode(mode) when mode == Vim.Types.Insert =>
+      Actions.checkCompletionMeet(state)
+
+    | ChangeMode(mode) when mode != Vim.Types.Insert => Actions.stop(state)
+
+    | EditorCursorMove(_) when state.mode == Vim.Types.Insert =>
+      Actions.checkCompletionMeet(state)
+
+    | CompletionAddItems(_meet, items) => {
+        let all = List.concat([items, state.completions.all]);
+        let filtered = Completions.filterItems(state.completions.filter, all);
+        Actions.update(
+          model =>
+            {
+              ...model,
+              all,
+              focused:
+                model.focused == None && Array.length(filtered) > 0
+                  ? Some(0) : model.focused,
+              filtered,
+            },
+          state,
+        );
+      }
+
```

Comment:
I'm not entirely sure about the use of an `Actions` module here. It certainly makes this `updater` much clearer, and compositions well with `checkCompletionMeet`, but I'm not sure it'll work well more generally where it might just add indirection for little benefit.

Reply:
It's an interesting experiment! Thanks for exploring ways of improving this code.